File: src/sensor/sensor_scanner.py
"""Receiving and processing data from Atlas Scientific sensors."""
import time
import serial

#from serial import SerialException, SerialTimeoutException



def store_ports():
    """Storing sensor ports."""
    ports_list = []

    ph = serial.Serial('/dev/ttyUSB0', 9600, timeout=1)
    ports_list.append(ph)
    do = serial.Serial('/dev/ttyUSB1', 9600, timeout=1)
    ports_list.append(do)
    orp = serial.Serial('/dev/ttyUSB2', 9600, timeout=1)
    ports_list.append(orp)
    ec = serial.Serial('/dev/ttyUSB3', 9600, timeout=1)
    ports_list.append(ec)
    rtd = serial.Serial('/dev/ttyUSB4', 9600, timeout=1)
    ports_list.append(rtd)

    return ports_list

def one_port_read():
    """This function can be used to test different ports."""

    dh = serial.Serial('dev/ttyUSB1', 9600, timeout=1)
    dh.write(b"R\r")

    time.sleep(1)
    while dh.in_waiting:
        data = dh.readline().decode().split()
        print(f"{dh.port}: {data}")

def single_read():
    """Get a single reading from the sensor to see if connection was successful."""
    ports = store_ports()
    serials = []

    for port in ports:
        try:
            serials.append(port)
        except serial.SerialException as s:
            print(f"Could not find or connect to port: {port}: {s}")

    while True:
        for srl in range(len(serials)): # check this line if it resolves our problem with the iteration
            srl_char = serials[srl].read(1)

            try:
                srl_char = srl_char.readline().decode().split()
            except UnicodeDecodeError:
                pass


if __name__ == "__main__":
    single_read()

File: src/sensor/sensor_scanner.cpp
#include <iostream>
#include <vector>
#include <fcntl.h>
#include <errno.h>
//#include <termios.h>
//#include <unistd.h> 

struct Ports {
    const char* sensor;
    const char* port;
};

auto get_ports() {
    Ports ports;

    ports.sensor = "ph";
    ports.port = "/dev/ttyUSB0";

    std::vector<const char*> vec_port = {ports.sensor, ports.port};

    for (const char* p: vec_port) {
        int serial_port;
        const char* serial_port_ptr = ports.sensor;
        serial_port = open(serial_port_ptr, O_RDWR);
        std::cout << serial_port;
    }
}

/*
auto read_ports() {
    std::vector<Ports> ports = {
        {"ph", "/dev/ttyUSB0"},
        {"od", "/dev/ttyUSB1"}
        
    };

    for(const char* p: ports){
       int serial_port;
       const char* serial_port_ptr = &p.port;
       serial_port = open(serial_port_ptr, O_RDWR);
    }
}
*/

int main() {
    get_ports();
}

